newrelic/hooks/mlmodel_openai.py

```diff
@@ -0,0 +1,145 @@
+
+# Copyright 2010 New Relic, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import openai
+import uuid
+from newrelic.api.function_trace import FunctionTrace
+from newrelic.common.object_wrapper import wrap_function_wrapper
+from newrelic.api.transaction import current_transaction
+from newrelic.api.time_trace import get_trace_linking_metadata
+from newrelic.core.config import global_settings
+from newrelic.common.object_names import callable_name
+from newrelic.core.attribute import MAX_LOG_MESSAGE_LENGTH
+
+
+def wrap_chat_completion_create(wrapped, instance, args, kwargs):
+    transaction = current_transaction()
+
+    if not transaction:
+        return
+
+    ft_name = callable_name(wrapped)
+    with FunctionTrace(ft_name) as ft:
+        response = wrapped(*args, **kwargs)
+
+    if not response:
+        return
+
+    custom_attrs_dict = transaction._custom_params
+    conversation_id = custom_attrs_dict.get("conversation_id", str(uuid.uuid4()))
```

Comment:
We decided this in a recent meeting and I've updated the CDD's accordingly since then. If the customer doesn't set the conversation id on the transaction then it should be defaulted to NULL. I think I might clarify this to say it won't be present if the user doesn't set it so this code should be:
```suggestion
    conversation_id = custom_attrs_dict.get("conversation_id", "")
```